### PR TITLE
Clone mutated inputs in first pass of CPP wrapper compilation

### PR DESCRIPTION
**Pull request by aakhundov**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123316

Summary: CPP wrapper compilation is currently done in two passes: in the first pass, Python wrapper is generated and run to compile Triton kernels as a side effect, in the second pass C++ wrapper is generated and compiled. When model inputs are mutated, running the Python wrapper in the first pass mutates the inputs, although the first pass (including the Python wrapper run) is strictly a part of the compilation process, hence must not introduce any side effects on the example inputs.

In this PR, we clone mutated inputs in the first pass to avoid input mutation.

Fixes https://github.com/pytorch/pytorch/issues/117364.

Test Plan:

```
$ TORCHINDUCTOR_CPP_WRAPPER=1 python test/inductor/test_torchinductor.py -k test_inductor_layout_optimization_input_mutations_cuda
...
.
----------------------------------------------------------------------
Ran 1 test in 6.368s

OK
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by desertfire — COMMENTED** [review]
Does input mutation changes the final output code?

**Comment by aakhundov** [comment]
> Does input mutation changes the final output code?

Not sure I understand the question correctly, but if you mean the input mutation that is prevented in this PR: no, it doesn't change the output code, just the numerics. E.g., when we do this with an input-mutating `f`:

```python
x = torch.rand(10, 20)
f_compiled = torch.compile(f, options={"cpp_wrapper": True})
y = f_compiled(x)
```

the result in `y` will be incorrect due to double mutation of `x`: first time during the Python wrapper run in the first phase, second time during the actual run of the compiled C++ wrapper code. Does this answer your question?

File: torch/_inductor/graph.py [hunk 1]
```diff
@@ -1285,6 +1285,19 @@ def materialize(x):
...
+                mutated_input_idxs = [
+                    idx
+                    for idx, name in enumerate(self.graph_inputs)
+                    if name in self.mutated_inputs
+                    and isinstance(real_inputs[idx], torch.Tensor)
+                ]
+                for idx in mutated_input_idxs:
+                    # clone mutated Tensor inputs to avoid mutating the them
+                    # in the first pass of the CPP wrapper compilation, as this
+                    # will lead to double mutation otherwise
```
Comment (chenyang78) [inline torch/_inductor/graph.py:1298]: Hmm, I have a silly question - what does double mutation mean here? Thanks!

**Comment by chenyang78** [comment]
Do we want to enable test_inductor_layout_optimization_input_mutations_cuda, assuming it's disable? Otherwise, it would fail? Thanks.

File: torch/_inductor/graph.py (hunk 1 above)
Comment (aakhundov) [reply torch/_inductor/graph.py:1298]: It's misleading in the comment, let me rephrase it. But what I mean with "double mutation" is summarized in the [previous comment](https://github.com/pytorch/pytorch/pull/123316#issuecomment-2035996966). Please let me know if that clarifies.

**Comment by aakhundov** [comment]
> Do we want to enable test_inductor_layout_optimization_input_mutations_cuda, assuming it's disable?

I'm actually not sure where to enable it. Grepping `layout_optimization_input_mutations` in the codebase only points at the `test_torchinductor.py`. Do we have a dedicated CI job for running tests from `test_torchinductor.py` with CPP wrapper?

File: torch/_inductor/graph.py (hunk 1 above)
Comment (aakhundov) [reply torch/_inductor/graph.py]: I've extended the comment, hope it reads more clear now.

File: torch/_inductor/graph.py [hunk 2]
```diff
@@ -1285,6 +1285,23 @@ def materialize(x):
...
+                ]
+                for idx in mutated_input_idxs:
+                    # clone mutated Tensor inputs to avoid mutating them in
+                    # the first pass of the CPP wrapper-based compilation, as
+                    # this will lead to a side effect on the example inputs:
+                    # e.g. if torch.compile(f)(x) if called on input-mutating
+                    # f, the inputs x will be mutated twice in the process:
+                    # once here, and again when running the compiled model;
+                    # this will also lead to a numerically incorrect output
+                    real_inputs[idx] = real_inputs[idx].clone()
```
Comment (jansel) [inline torch/_inductor/graph.py:1303]: I think this needs to be `clone_preserve_strides`

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/graph.py (hunk 1 above)
Comment (chenyang78) [reply torch/_inductor/graph.py]: I see. Thanks for the clarification.

**Review by chenyang78 — APPROVED** [review]
LGTM. Thanks!

**Review by desertfire — APPROVED** [review]
Approved pending test to be added.

**Review by eellison — COMMENTED** [review]
For things like in-placing updating parameters in the optimizer, cloning will have a bad memory effect. Running the backward while holding the activations in memory will also hurt memory. 

I agree we should land this but for longer-term compatibility with training I think we'll need to figure out a different solution than two-pass on first invocation.

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
Please add a test

**Comment by aakhundov** [comment]
Following the suggestion, added `test_inductor_layout_optimization_input_mutations` to `test_cpu_cpp_wrapper.py` and `test_cuda_cpp_wrapper.py`.
